plugins/core/src/main/kotlin/de/fayard/refreshVersions/core/DependencyGroup.kt

```diff
@@ -0,0 +1,172 @@
+    @InternalRefreshVersionsApi
+    var usePlatformConstraints: Boolean = false
+) {
+
+    private val rule: ArtifactVersionKeyRule? = rawRule?.let {
+        val lines = it.lines()
+        assert(lines.size == 2) {
+            "2 lines were expected, but ${lines.size} were found: $it"
+        }
+        ArtifactVersionKeyRule(
+            artifactPattern = lines.first(),
+            versionKeyPattern = lines.last()
+        )
+    }
+
+
+    companion object {
+        private val ALL = mutableListOf<AbstractDependencyGroup>()
+
+        @InternalRefreshVersionsApi
+        val ALL_RULES: List<ArtifactVersionKeyRule>
+            get() = ALL.mapNotNull { it.rule }
+
+        @InternalRefreshVersionsApi
+        var disableBomCheck: Boolean = false
+
+        private val isRunningTests: Boolean by lazy {
+            try {
+                Class.forName("org.junit.jupiter.api.AssertEquals")
+                true
+            } catch (e: ClassNotFoundException) {
+                false
+            }
+        }
+    }
+
+    init {
+        assert(group.isNotBlank()) { "Group shall not be blank" }
+        ALL.add(this)
+    }
+
+    fun module(
+        name: String,
+        isBom: Boolean = false,
+        usePlatformConstraints: Boolean? = if (isBom) false else null
+    ): DependencyNotation {
+        assert(name.trimStart() == name) { "module($name) has superfluous leading whitespace" }
+        assert(name.trimEnd() == name) { "module($name) has superfluous trailing whitespace" }
+        assert(name.contains(":").not()) { "module($name) is invalid" }
+        return object : DependencyNotationImpl(
+            group = group,
+            name = name
+        ) {
+
+            override fun shouldUsePlatformConstraints(): Boolean {
+                markDependencyNotationsUsage()
+                return usePlatformConstraints ?: this@AbstractDependencyGroup.usePlatformConstraints
+            }
+
+            private fun markDependencyNotationsUsage() {
+                if (isRunningTests || disableBomCheck) return
+
+                val isGroupUsingPlatformConstraints = this@AbstractDependencyGroup.usePlatformConstraints
+
+                if (isBom) {
+                    if (isGroupUsingPlatformConstraints.not() &&
+                        haveDependencyNotationsWithPlatformConstraintsBeenUsed
+                    ) {
+                        error(
+                            "You are trying to use a BoM ($artifactPrefix), " +
+                                "but dependency notations relying on it have been declared before! " +
+                                "Declare the BoM first to fix this issue."
+                        )
+                    }
+                    this@AbstractDependencyGroup.usePlatformConstraints = true
+                } else if (isGroupUsingPlatformConstraints) {
+                    haveDependencyNotationsWithPlatformConstraintsBeenUsed = true
+                }
+            }
```

Comment:
Please, double check that the boolean logic is sound

Reply:
I think the issue is that the naming `usePlatformConstraints` is not very easy to hold in mind as to what it actually means.

I'll try to find something that is easier to reason about and check that logic.

Reply:
Resolved on main